### PR TITLE
Refactored DatabaseErrorDialog to MaterialAlertDialog

### LINKED ISSUE #11780: deprecation: Material Dialog neutral button
Please log an issue to track this

_Originally posted by @mikehardy in https://github.com/ankidroid/Anki-Android/pull/11646#discussion_r913977376_

### PR DESCRIPTION
**Pull request by neeldoshii**
## Fixes
* Related https://github.com/ankidroid/Anki-Android/issues/13315
* Fixes #11780

## Screenshots

`INCOMPATIBLE_DB_VERSION`

| Before | After |
|--------|--------|
| screenshot | screenshot |

`DIALOG_DB_ERROR`

| Before | After |
|--------|--------|
| screenshot | screenshot |

`DIALOG_RESTORE_BACKUP`

| Before | After |
|--------|--------|
| screenshot | screenshot |

`DIALOG_STORAGE_UNAVAILABLE_AFTER_UNINSTALL`
| Before | After |
|--------|--------|
| screenshot | screenshot |

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 1]
```diff
@@ -53,17 +51,14 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
...
-    @Suppress("Deprecation") // Material dialog neutral button deprecation
     @SuppressLint("CheckResult")
     override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
         super.onCreateDialog(savedInstanceState)
         val res = res()
         val alertDialog = AlertDialog.Builder(requireActivity())
         val dialog = MaterialDialog(requireActivity())
+        val materialAlertDialog = MaterialAlertDialogBuilder(requireActivity())
         val isLoggedIn = isLoggedIn()
-        dialog.cancelable(true)
```
Comment (neeldoshii) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:64]: is this okay to remove I felt this is redundant.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 2]
```diff
@@ -53,17 +51,14 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
     private lateinit var repairValues: IntArray
     private lateinit var backups: Array<File>
 
-    @Suppress("Deprecation") // Material dialog neutral button deprecation
```
Comment (neeldoshii) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:56]: Successful Merge of this PR will close #11780

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 3]
```diff
@@ -361,6 +357,7 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
             }
             DIALOG_STORAGE_UNAVAILABLE_AFTER_UNINSTALL -> {
                 val listItems = UninstallListItem.createList()
+                // AlertDialog does not allow message and listItems
```
Comment (neeldoshii) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:360]: Not sure in materialAlertDialog also over here `listItem` & `message` is not working.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 4]
```diff
@@ -519,22 +516,6 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
             }
             else -> requireArguments().getString("dialogMessage")
         }
-    private val title: String
```
Comment (neeldoshii) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:522]: title is now unused.

**Comment by neeldoshii** [comment]
@david-allison Should I create a new PR for this commit 057ea62f5ce3a93f78044000d12a98de9f653300 as this is functional change in refactor PR as well its blocking #16266

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 2 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:56]: Added to description

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:64]: why? `DIALOG_ERROR_HANDLING`, `DIALOG_REPAIR_COLLECTION` etc... seems to need it

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 5]
```diff
@@ -331,14 +319,14 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                     options.add(makeBold(res.getString(R.string.backup_one_way_sync_from_server)))
                     values.add(1)
                 }
-                dialog.show {
+                alertDialog.show {
                     cancelable(false)
-                    contentNullable(message)
-                    icon(R.drawable.ic_warning)
+                    title(R.string.incompatible_database_version_title)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:324]: This looks wrong

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 6]
```diff
@@ -361,16 +349,15 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
             }
             DIALOG_STORAGE_UNAVAILABLE_AFTER_UNINSTALL -> {
                 val listItems = UninstallListItem.createList()
-                dialog.show {
-                    contentNullable(message)
-                    listItems(items = listItems.map { getString(it.stringRes) }, waitForPositiveButton = false) { _: MaterialDialog, index: Int, _: CharSequence ->
+                alertDialog.show {
+                    title(R.string.directory_inaccessible_after_uninstall)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:353]: why is `message` no longer being set?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 7]
```diff
@@ -214,14 +207,14 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                     val formatter = LocalizedUnambiguousBackupTimeFormatter()
                     val dates = backups.map { formatter.getTimeOfBackupAsText(it) }
 
-                    dialog.title(R.string.backup_restore_select_title)
+                    alertDialog.title(R.string.backup_restore_select_title)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:210]: keep it as `dialog` to reduce lines in the diff

File: AnkiDroid/src/main/java/com/ichi2/utils/AlertDialogFacade.kt [hunk 8]
```diff
@@ -328,3 +331,25 @@ fun AlertDialog.Builder.listItems(items: List<CharSequence>, onClick: (dialog: D
...
+
+/**
+ * Extension workaround for Displaying ListView & Message Together
+ * Alert Dialog Doesn't allow message and listview together so a customView is used.
+ *
+ * @param messageResID The message which you want to display in the dialog
+ * @param items The items to display in the list.
+ * @param onClick A lambda function that is invoked when an item is clicked.
+ */
+fun AlertDialog.Builder.listItemsAndMessage(@StringRes messageResId: String?, items: List<CharSequence>, onClick: (dialog: DialogInterface, index: Int) -> Unit): AlertDialog.Builder {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/utils/AlertDialogFacade.kt:343]: A `string` isn't a `StringRes` or a `resId`, it's

`message: String?`

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 4 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:522]: I'd have preferred to keep a minimal diff

File: AnkiDroid/src/main/res/layout/dialog_listview_message.xml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:padding="3dp"
+    android:orientation="vertical">
+
+    <!--Dialog Message-->
+    <TextView
```
Comment (david-allison) [inline AnkiDroid/src/main/res/layout/dialog_listview_message.xml:27]: `FixedTextView`

**Review by david-allison — CHANGES_REQUESTED** [review]
Could you post before & after screenshots, `INCOMPATIBLE_DB_VERSION` looks wrong.

I feel the styling may be a little worse in the new version

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 4 above)
Comment (neeldoshii) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:522]: Wont this cause lint test failing?  'variable unused'

**Comment by neeldoshii** [comment]
I am on break due to university exams, drafting this PR till then. Will open it for review once the `requested changes` are refactored for review

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 1 above)
Comment (neeldoshii) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:64]: fixed.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 6 above)
Comment (neeldoshii) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:355]: Thanks for the catch. fixed it.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 4 above)
Comment (neeldoshii) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:522]: done.

File: AnkiDroid/src/main/java/com/ichi2/utils/AlertDialogFacade.kt (hunk 8 above)
Comment (neeldoshii) [reply AnkiDroid/src/main/java/com/ichi2/utils/AlertDialogFacade.kt]: fixed.

**Comment by neeldoshii** [comment]
Screenshots of `INCOMPATIBLE_DB_VERSION`. Note : Ignore Before's background dim its emulator issue.

I feel the styling may be a little worse in the new version
| Before | After |
|--------|--------|
| screenshot | screenshot |

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt (hunk 7 above)
Comment (neeldoshii) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:212]: I don't think its possible we are already using `alertdialog` variable for alertdialogs. Introduced in #15396

**Comment by neeldoshii** [comment]
@david-allison Is this because of my fast commits pushed?

```
  Received 1440578237 of 1440578237 (100.0%), 85.6 MBs/sec
  Cache restored successfully
  Restored groovy-dsl with key gradle-groovy-dsl-v1-b6176636e15c884e52aa481f4216903e to C:\Users\runneradmin\.gradle\caches\*\groovy-dsl\*/
  Error: The action 'Setup Gradle' has timed out after 5 minutes.
  ```

**Comment by david-allison** [comment]
Don't worry about CI, we're having issues

**Comment by david-allison** [comment]
My request would be to have side-by-side images for each screen.

A few of them look unusual, and I want to confirm visually there's no regressions.

Once we have a list of 'before' screenshots, they won't need to change

**Review by david-allison — COMMENTED** [review]
The "incompatible DB version" in "Before" looks incorrect

From the code on `main`, it looks like it should be using `incompatible_database_version_summary`, can you confirm why this isn't the case?

**Comment by neeldoshii** [comment]
Updated all the screenshots in the PR description for review.


> The "incompatible DB version" in "Before" looks incorrect
> 
> From the code on main, it looks like it should be using incompatible_database_version_title & incompatible_database_version_summary, can you confirm why this isn't the case?

Its because I did  manual trigger for testing at `DIALOG_CONFIRM_DATABASE_CHECK` so its taking its corresponding title & message rather than `incompatible_database_version_title` & `incompatible_database_version_summary`.
```
            DIALOG_CONFIRM_DATABASE_CHECK -> {
                // Confirmation dialog for database check
                alertDialog.show {
                    title(R.string.check_db_title)
                    message(text = message)
                    positiveButton(R.string.dialog_ok) {
                        (activity as DeckPicker).integrityCheck()
                        dismissAllDialogFragments()
                    }
                    negativeButton(R.string.dialog_cancel)
                }
            }
```

**Comment by david-allison** [comment]
Neither the before nor the 'after' screenshots seem correct, can we do better here?

**Comment by neeldoshii** [comment]
updated the screenshots for review ;)

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt [hunk 10]
```diff
@@ -214,14 +209,14 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
...
                         .positiveButton(R.string.restore_backup_choose_another) {
                             (activity as? AnkiActivity)?.let {
                                 ImportFileSelectionFragment.openImportFilePicker(it, ImportFileSelectionFragment.ImportFileType.APKG)
                             }
                         }
                         .negativeButton(R.string.dialog_cancel)
-                        .listItemsSingleChoice(items = dates.toTypedArray().toList(), waitForPositiveButton = false) { _: MaterialDialog, index: Int, _: CharSequence ->
+                        .setSingleChoiceItems(dates.toTypedArray(), -1) { _, index: Int ->
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:219]: From the docs:
`If -1 no items are checked.`

**Review by david-allison — APPROVED** [review]
Looks great, thanks!

**Review by BrayanDSO — APPROVED** [review]
LGTM. Thanks

**Pull request merged**